File: backend/app/schemas/appointment.py
"""Scheduling DTOs (appointments calendar)."""
from __future__ import annotations

from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

AppointmentStatus = Literal["booked", "arrived", "done", "cancelled", "no_show"]


class AppointmentOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    appointment_no: str
    branch_id: UUID
    patient_id: UUID
    patient_name: str
    doctor_id: UUID | None
    doctor_name: str | None = None
    cabinet: str | None
    service: str | None
    starts_at: datetime
    ends_at: datetime
    status: str
    notes: str | None
    created_at: datetime


class AppointmentCreate(BaseModel):
    branch_id: UUID
    patient_id: UUID
    doctor_id: UUID | None = None
    starts_at: datetime
    duration_min: int = Field(default=30, ge=5, le=240)
    cabinet: str | None = Field(None, max_length=32)
    service: str | None = Field(None, max_length=255)
    notes: str | None = Field(None, max_length=512)


class AppointmentReschedule(BaseModel):
    starts_at: datetime
    duration_min: int | None = Field(None, ge=5, le=240)
    doctor_id: UUID | None = None
    cabinet: str | None = Field(None, max_length=32)


class AppointmentStatusUpdate(BaseModel):
    status: AppointmentStatus


class SchedStaffOut(BaseModel):
    """Doctor/staff member for the calendar columns (under appointments.read)."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    full_name: str
    roles: list[str] = []

File: lib/features/scheduling/domain/appointment.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'appointment.freezed.dart';
part 'appointment.g.dart';

/// A booked appointment (mirrors backend `AppointmentOut`).
/// Status flow: booked → arrived → done | cancelled | no_show.
@freezed
abstract class Appointment with _$Appointment {
  const Appointment._();

  const factory Appointment({
    required String id,
    required String appointmentNo,
    required String branchId,
    required String patientId,
    @Default('') String patientName,
    String? doctorId,
    String? doctorName,
    String? cabinet,
    String? service,
    required String startsAt,
    required String endsAt,
    required String status,
    String? notes,
    required String createdAt,
  }) = _Appointment;

  factory Appointment.fromJson(Map<String, dynamic> json) =>
      _$AppointmentFromJson(json);

  /// Local start time (the backend serialises aware-UTC). A naive string is
  /// treated as UTC before converting to local.
  DateTime? get start => _local(startsAt);
  DateTime? get end => _local(endsAt);

  static DateTime? _local(String s) {
    if (s.isEmpty) return null;
    final hasZone = s.endsWith('Z') || RegExp(r'[+-]\d\d:?\d\d$').hasMatch(s);
    return DateTime.tryParse(hasZone ? s : '${s}Z')?.toLocal();
  }

  String get timeLabel {
    final d = start;
    if (d == null) return '';
    return '${d.hour.toString().padLeft(2, '0')}:${d.minute.toString().padLeft(2, '0')}';
  }

  String get statusLabel => switch (status) {
        'booked' => 'Записан',
        'arrived' => 'Пришёл',
        'done' => 'Принят',
        'cancelled' => 'Отменён',
        'no_show' => 'Не пришёл',
        _ => status,
      };
}
